code/game/machinery/doors/firedoor.dm

```diff
@@ -78,17 +82,153 @@
 	. = ..()
 	INVOKE_ASYNC(src, .proc/latetoggle)
 
-/obj/machinery/door/firedoor/attack_hand(mob/user, list/modifiers)
+/obj/machinery/door/firedoor/attack_hand(mob/living/user, list/modifiers)
 	. = ..()
 	if(.)
 		return
 	if(operating || !density)
 		return
-	user.changeNext_move(CLICK_CD_MELEE)
+	if(user.combat_mode)
+		user.changeNext_move(CLICK_CD_MELEE)
 
-	user.visible_message("<span class='notice'>[user] bangs on \the [src].</span>", \
-		"<span class='notice'>You bang on \the [src].</span>")
-	playsound(loc, 'sound/effects/glassknock.ogg', 10, FALSE, frequency = 32000)
+		user.visible_message("<span class='notice'>[user] bangs on \the [src].</span>", \
+			"<span class='notice'>You bang on \the [src].</span>")
+		playsound(loc, 'sound/effects/glassknock.ogg', 10, FALSE, frequency = 32000)
+	else
+		if(user.do_afters)//you get one firelock to open, better pick the right one
```

Comment:
Use `DOING_INTERACTION`, and create an explicit interaction key for firelocks. That way, you limit their focus to the firelocks, but you do not disrupt any other potential do_afters they have (including potentially abstract ones).